Header | Icon | Button 컴포넌트 추가

Rengod95 opened the pull request:
작업 이슈

- 헤더 컴포넌트를 추가
- CSS Reset 코드를 추가
- 임시 컬러 프로필 추가
- Button.tsx 추가
- Icon.tsx 추가
- FlexBoxCreator 추가

작업 상세 설명
https://github.com/Rengod95/Dynamic_Roadmap/pull/3/commits/ef3395ce124954a8318ab046e75bfb71e0ff6b2e
[f4b855a](https://github.com/Rengod95/Dynamic_Roadmap/pull/3/commits/f4b855af2cf951b0fcb451c4fefd31c5d823e5a1)
[80db2c0](https://github.com/Rengod95/Dynamic_Roadmap/pull/3/commits/80db2c07928d205cfa972f59d106a14f7cfedea0)

-- PR 코멘트 반영 -- 

https://github.com/Rengod95/Dynamic_Roadmap/pull/3/commits/776afc7dd727f1f8a838a8131e403a3f31180252

Rengod95 committed:
feat: Add Header Component
헤더 컴포넌트를 추가했습니다.

기본적인 CSS Reset 코드를 추가했습니다.
임시 컬러 프로필 scss를 추가했습니다.

components 폴더의 구조를 변경했습니다.

Rengod95 committed:
fix: refactor Header.tsx

자주 사용하는 flex 스타일링이 적용된 컴포넌트를 간편하게 생성할 수 있는
createFlexBox.ts 추가

styled 의 인자로 사용하여 스타일링 오버라이드 가능하도록 설계

Header.tsx를 구성하는 각 styled component에 대한 이름 변경

Rengod95 committed:
feat: Initialize React-query

* react-quert-devtools 추가
* next.js 에서 react-query를 SSR 및 SSG로 사용하기 위한 보일러플레이트 추가
  - lazy init을 통해 하나의 quertClient 사용
* PostList.tsx PostListNav.tsx Post.tsx 추가
 - PostList : 데이터 페칭을 통해 게시글 목록을 출력
 - Post : 각 게시글 디테일 컴포넌트
 - PostListNav : PostList의 출력 리스트를 페이지네이션 할 수 있도록 하는 네비게이션 컴포넌트

 * Header.tsx 를 구성하는 스타일링 컴포넌트의 이름 수정
  - 각 컴포넌트의 역할을 가시적으로 확인 가능하도록 변경

* FlexListitem.tsx 컴포넌트 추가
 - 자주 사용하는 li 컴포넌트에 대해 Flex css 가 적용된 추상화된 컴포넌트로 기능

Rengod95 committed:
feat: Add Common Components

* Button.tsx 컴포넌트 추가
 - size와 variant에 따라 크기 및 컬러 프리셋 설정
 - 임의의 사이즈 및 컬러 설정을 위해 타입을 분할해서 선언
   -ColorPreset | Color  SizePreset | Size
 - title, onClick, disbabled, children 을 받을수 있도록 추상화
* Icon.tsx 컴포넌트 추가
 - ICON_SET.ts 을 통해 아이콘 이름과, path만 추가해도 간단하게 아이콘 추가 가능하도록 설정
 - viewPortSize와 viewBoxSize를 다양한 형태로 기입할 수 있도록 인자를 리터럴 타입으로 설정

 * createFlexBox 수정
  - FlexBoxCreator 클래스 추가
  - 싱글턴으로 단일 FlexBoxCreator를 참조하도록 설정
  - style 작성시 Flex css 작성하는 부분을 간소화 하기 위한 제작
  - styled(FlexBoxCreator.createDefaultFlexBox(컴포넌트 | 스트링)``  형태로 사용 가능
  - StyledComponent 반환, 합성 가능

imnotmoon commented on package.json:
```diff
@@ -22,6 +24,7 @@
     "next": "13.1.6",
     "react": "18.2.0",
     "react-dom": "18.2.0",
+    "sass": "^1.58.0",
```
잘 몰라서 그러는데 sass 의존성이 왜 필요한가요?

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+import React from 'react';
+import { FlexListItem, FlexBoxCreator } from '../layout/FlexBox';
+
+type PostListProps = {
```
모든 타입은 export 해주세요.
밖에서 쓸지 안쓸지 모르겠다 싶은건 하는게 맞습니다.

Rengod95 replied:
넵 수정했습니다

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+type Posts = { posts: Post[] };
+
+export const PostList = () => {
+  const { data, isError, isLoading, isSuccess } = useQuery(['posts'], () => getPosts());
```
useQuery 훅은 PostList.hooks.ts에 만들고 import 해주세요.
컴포넌트 파일에서 컴포넌트 로직에 집중하기 어렵습니다.

imnotmoon replied:
```ts
// PostList.hooks.ts
export const usePosts = (options) => useQuery(['posts'], getPosts, options)
```

imnotmoon replied:
생각해보니 query key는 단수 표현으로 써야합니다. ['post']가 되겠네요.

Rengod95 replied:
데이터 페칭 로직을 PostList.tsx 컴포넌트에서 /Pages/board/post/[id]/index.tsx 로 이동했습니다!
쿼리 키 부분 수정 했습니다

Rengod95 replied:
다시 PostList로 이동해 훅으로 분리했습니다!

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+export const PostList = () => {
+  const { data, isError, isLoading, isSuccess } = useQuery(['posts'], () => getPosts());
+
+  if (isSuccess) {
+    return (
+      <PostListContainer>
+        <StyledPostList>
+          {data.posts.map((post) => {
+            return <PostListItem key={post.id}>{post.id}</PostListItem>;
+          })}
+        </StyledPostList>
+      </PostListContainer>
+    );
+  }
+  console.log(data);
+  return <div></div>;
```
- 삼항연산자 이용해서 return문을 하나로 만들어주세요.
- console.log는 PR에 올라가면 안됩니다. 저도 가끔 실수하는데 빼주시는게 좋아요.

Rengod95 replied:
삼항연산자보다 타입스크립트 문법인 && 이 더 깔끔한것 같아 이 방식으로 수정 했는데 괜찮을까요?

imnotmoon replied:
네 &&도 전혀 안티패턴 아니고 자주 쓰입니다.
근데 생각해보니 isSuccess라고 굳이 표현할 필요 없을것 같아요. 컴포넌트에서는 해피 케이스만 다루는게 최선이긴 합니다.
오히려 에러케이스를 어떻게 처리할지에 대해서 고민해주세요.

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+  dehydratedState: ReturnType<typeof dehydrate>;
+};
+
+type Post = {
```
PostListItem이나 hook 파일에 선언되는게 더 적절할것 같습니다.

+ 생각해보니 Post 타입을 아직 우리는 알 수 없습니다. 차라리 정의를 안하는게 낫습니다. 꼭 할거면 id정도만 해주세요.

Rengod95 replied:
넵 저도 각 개체의 속성을 임의로 정의하는 것은 불필요한 코드 작성이라고 판단했습니다.

현재는 타입스크립트 학습 + 더미 API를 이용해서 정상적으로 출력되는지 테스트 하기 위해, 더미 API가 반환해주는 데이터를 기준으로 interface를 작성해놓은 상태입니다.

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+  tags: string[];
+};
+
+type Posts = { posts: Post[] };
```
Posts는 Post의 복수 표현이라 Post[]일 것으로 예상되는데 실제로는 { posts: Post[] }입니다.
어색한 코드예요.

Rengod95 replied:
데이터 스키마에 끼워맞추다 보니 말씀해 주신대로 어색한 타입 선언이 되어버렸습니다. 수정하도록 하겠습니다!

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+    return (
+      <PostListContainer>
+        <StyledPostList>
+          {data.posts.map((post) => {
```
data가 undefined일 경우 터질 수 있는 위험한 코드입니다.
data를 구조분해 해서 기본값을 주세요.

Rengod95 replied:
넵! data를 리턴 이전에 Destructuring 하는 방식으로 변경하겠습니다

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+
+  if (isSuccess) {
+    return (
+      <PostListContainer>
```
Container는 여백을 잡기 위해 사용하는 컴포넌트로 어느정도 관습이 있습니다.
이 경우에는 어울리지 않는 이름입니다.

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+  if (isSuccess) {
+    return (
+      <PostListContainer>
+        <StyledPostList>
```
Styled가 꼭 붙어야 하나요?

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+      dehydratedState: dehydrate(queryClient),
+    },
+  };
+};
```
이 부분은 제 공부가 덜 돼서 이후에 다시 확인해볼게요

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+  };
+};
+
+export const getPosts = async (offset: number = 0): Promise<Posts> => {
```
Posts를 response 용도로 쓸거면 PostsResponse가 더 적당한 이름일것 같습니다.

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+};
+
+export const getPosts = async (offset: number = 0): Promise<Posts> => {
+  return await fetch(`https://dummyjson.com/posts`).then((res) => res.json());
```
항상 .then(res => res.json())이 필요하면 fetch를 데코레이팅한 함수를 만들어도 좋겠어요.

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+  return await fetch(`https://dummyjson.com/posts`).then((res) => res.json());
+};
+
+const PostListContainer = styled(FlexBoxCreator.columnStartFlexBox('section'))`
+  display: block;
+  width: 1080px;
+  min-height: 720px;
+  padding: 16px;
+  margin: 0 auto;
+`;
+
+const StyledPostList = styled(FlexBoxCreator.columnStartFlexBox('ul'))`
+  width: 100%;
+  height: 100%;
+  gap: 1rem;
+`;
+
+const PostListItem = styled(FlexListItem)`
+  min-width: 1000px;
+  background-color: #424890;
+  border-radius: 5px;
+  width: 100%;
+  padding: 16px;
+  height: 60px;
+  flex-wrap: wrap;
+`;
```
스타일 파일은 PostList.css.ts에 분리해서 작성해주세요.
마찬가지로 컴포넌트 코드에 대한 집중을 분산시킵니다.

Rengod95 replied:
넵 반영하겠습니다! emotion이나 StyledComponent 등을 이용해서 작성된 단순 스타일링이 적용된 컴포넌트의 경우 분리해서 작성한다고 이해해도 되는 부분일까요?

imnotmoon replied:
네 컴포넌트는 뷰만 있는게 최선입니다.
로직은 어쩔 수 없이 섞이긴 하지만 가능한 한 뷰와 분리되었으면 좋겠어요.
스타일은 같은 파일에 위치시킬 이유가 없습니다. 시선만 분산시켜요.

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+};
+
+const PostListContainer = styled(FlexBoxCreator.columnStartFlexBox('section'))`
+  display: block;
+  width: 1080px;
+  min-height: 720px;
+  padding: 16px;
+  margin: 0 auto;
```
- section은 원래 block content 아닌가요?
- 고정 width와 min-height가 벌써 필요한 이유가 궁금해요.
- padding이나 margin은 컴포넌트 루트에 잘 선언하지 않습니다. 변경에 취약해져요.

Rengod95 replied:
고정 width외 min-height의 경우 다양한 커뮤니티들이 리스트 아이템을 뿌려주는 컨테이너의 기본 사이즈를 다 고정 픽셀로 지정하여 작성하길래 참고해서 작성했습니다.

section 부분의 block은 말씀대로 수정하도록 하겠습니다.

padding, margin 부분도 수정하겠습니다. 다만 그렇다면 주로 어느 영역에서 margin과 padding을 설정하는지 알 수 있을까요?

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+const StyledPostList = styled(FlexBoxCreator.columnStartFlexBox('ul'))`
+  width: 100%;
+  height: 100%;
+  gap: 1rem;
```
flex-gap은 아직 시기상조입니다. 다른 방법으로 간격을 잡아보세요.
<img width="1382" alt="image" src="https://user-images.githubusercontent.com/53335940/219868779-16a0690e-0ba6-4190-8f1e-e040639f57bb.png">

Rengod95 replied:
스펙을 고려하지 않은 채 편한 코드로 작성하고 있었습니다.
탑 다운 형식으로 아이템들을 뿌려줘야 하는 플렉스 박스라 space-between 으로 대체하겠습니다

imnotmoon commented on src/components/Post/PostList.tsx:
```diff
@@ -0,0 +1,77 @@
+  gap: 1rem;
+`;
+
+const PostListItem = styled(FlexListItem)`
+  min-width: 1000px;
+  background-color: #424890;
+  border-radius: 5px;
+  width: 100%;
+  padding: 16px;
+  height: 60px;
+  flex-wrap: wrap;
```
일반적인 리스트 구조가 아니고 flex가 필요한 이유가 있나요?

Rengod95 replied:
현재 PostListItem을 생성할 때 파라미터로 전달되는 element는 Flex가 적용된 li 입니다.

단순히 생각했을 때, PostListItem 이 출력을 해 줘야하는 데이터가 정해지지 않는 상태에서 텍스트를 일관적으로 center에 배치하는 것이 확장성이 좋을 수 있겠다고 판단했습니다.

불필요한 코드라고 생각되시면 바로 수정하겠습니다!

imnotmoon commented on src/components/Post/PostListNav.tsx:
```diff
@@ -0,0 +1,52 @@
+import { FlexBoxCreator, FlexListItem } from '../layout/FlexBox';
+import styled from '@emotion/styled';
+import SVGIcon from '../commons/Icon/Icon';
+import Button from '../commons/Button/Button';
```
디자인이 어떻게 나올지 아직 몰라요. 이런 컴포넌트들이 왜 필요한지 설득해주세요.

imnotmoon commented on src/components/Post/PostListNav.tsx:
```diff
@@ -0,0 +1,52 @@
+
+export const PostListNav = () => {
+  return (
+    <PostNavContainer>
+      <PostNavLayout>
+        <PostNavItemContainer>
```
Container라는 이름은 꼭 필요한 경우가 아니면 쓰지 않기로 합시다.
Layout이라는 이름도 여기서 왜 필요한지 잘 이해하지 못했어요.

가이드 코드를 작성해서 PR로 올려볼테니 그거에 맞게 수정해주세요.

Rengod95 replied:
Container 나 Layout과 같은 이름은 제가 임의대로 작성했습니다. 

스타일링 컴포넌트의 네이밍이나 기본적으로 컴포넌트를 디자인할 때 스타일링 컨벤션, 네이밍 컨벤션 등에 대한 지식이 없어서 발생한 문제같습니다.

말씀해 주신대로 가이드 코드 보고 수정 해 보겠습니다 감사합니다!!

imnotmoon commented on src/components/Post/PostListNav.tsx:
```diff
@@ -0,0 +1,52 @@
+          </Button>
+        </PostNavItemContainer>
+        <PostNavItemContainer>
+          <PostNavNumberList></PostNavNumberList>
```
self-closing이 자동으로 적용되도록 하는 prettier 옵션을 찾아서 적용해주세요.

Rengod95 replied:
반영하도록 하겠습니다!

Rengod95 replied:
검색 결과 Prettier는 철학적인 이유로 self-closing과 관련된 옵션을 제공해 주지 않는다고 합니다. 
대신 IDE 자체의 포맷을 설정하는 방법이 있었습니다.
https://hello-sunil.in/stop-prettier-self-closing-tag/

imnotmoon commented on src/components/Post/PostListNav.tsx:
```diff
@@ -0,0 +1,52 @@
+  );
+};
+
+const PostNavContainer = styled(FlexBoxCreator.createDefaultFlexBox('div'))`
+  flex-wrap: nowrap;
+  width: 1080px;
+  height: 80px;
+  margin: 0 auto;
+`;
```
마찬가지로 flex-wrap이 왜 필요한지 모르겠고, margin은 여기 쓰면 안됩니다.
제 생각에 필요한건 height 정도밖에 없는것 같아요. 다 지워주세요.

Rengod95 replied:
넵 수정하겠습니다!!

imnotmoon commented on src/components/Post/PostListNav.tsx:
```diff
@@ -0,0 +1,52 @@
+  margin: 0 auto;
+`;
+const PostNavLayout = styled('div')`
+  display: grid;
+  grid-template-columns: 1.5fr auto 1.5fr;
+  grid-template-rows: 1fr;
```
grid 쓰지 않고 flex로만 처리해주세요.

Rengod95 replied:
넵 flex로 변경하도록 하겠습니다. 다만 grid를 쓰지 말아야 하는 이유가 따로 있을까요??

imnotmoon commented on src/components/commons/Button/BUTTON_PRESET.ts:
```diff
@@ -0,0 +1,29 @@
+import { ButtonSizePreset, ButtonSize, ButtonColor, ButtonColorPreset } from '.';
+
+export const PRESET_BUTTON_SIZES: Record<ButtonSizePreset, ButtonSize> = {
```
디자인이 나온건가요?

Rengod95 replied:
아닙니다. 디자인이 나왔을 때, PRESET 타입과 각 타입에 데이터만 추가하면 바로 사용할 수 있도록 작성했습니다.

imnotmoon commented on src/components/commons/Button/Button.tsx:
```diff
@@ -0,0 +1,43 @@
+  getButtonSize,
+} from '.';
+
+interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
```
interface는 되도록 쓰지 않으려고 해요. type 위주로 쓰고 extends가 필요한 경우는 &로 해결해주세요.

imnotmoon replied:
React의 Button 관련된 타입을 intersect 하면 onClick, disabled같은 타입 정의가 필요 없어집니다.

Rengod95 replied:
오오 또 하나 얻어가는 것 같습니다. 말씀해주신대로 변경 해보겠습니다.

다만 type vs interface 에서 interface를 지양하는 이유가 따로 있나요?

imnotmoon commented on src/components/commons/Button/Button.tsx:
```diff
@@ -0,0 +1,43 @@
+  onClick?: (...args: any) => any;
+  title?: string | undefined;
+  disabled?: boolean;
+  children?: ReactNode;
```
children을 prop으로 받으면 title이 필요 없겠죠?

imnotmoon commented on src/components/commons/Button/Button.tsx:
```diff
@@ -0,0 +1,43 @@
+  children?: ReactNode;
+}
+
+const Button = ({ size, variant, onClick, title, disabled, children }: ButtonProps) => {
```
Button이 너무 똑똑해지려는것 같아요. 똑똑해진다는게 좋게 들릴수도 있지만 장단점이 있어요.
사용하기 간단할 수는 있지만 학습을 위한 비용이 커지고, 변화에 취약해진다는 단점이 있습니다.

무엇보다도, 아직은 이런 작업이 필요 없을것 같아요.

imnotmoon commented on src/components/commons/Button/Button.tsx:
```diff
@@ -0,0 +1,43 @@
+
+const Button = ({ size, variant, onClick, title, disabled, children }: ButtonProps) => {
+  return (
+    <StyledButton onClick={onClick} disabled={disabled} size={size} variant={variant}>
```
```suggestion
    <StyledButton {...props}>
```
이렇게 간단하게 쓰면 더 좋겠어요.

imnotmoon commented on src/components/commons/Button/Button.tsx:
```diff
@@ -0,0 +1,43 @@
+  );
+};
+
+const StyledButton = styled(FlexBoxCreator.createDefaultFlexBox('button'))`
```
아무리 생각해도 지금은 필요 없습니다.
코드를 놔두되 사용하지 않거나, 삭제하는편이 좋겠어요.

imnotmoon commented on src/components/commons/Button/Button.tsx:
```diff
@@ -0,0 +1,43 @@
+  }}
+`;
+
+export default Button;
```
default export는 하지 않으려 합니다.

imnotmoon commented on src/components/commons/Button/button.util.ts:
```diff
@@ -0,0 +1,54 @@
+  PRESET_BUTTON_SIZES,
+} from '.';
+
+export const getButtonColor = (color: ButtonColor | ButtonColorPreset) => {
```
사용하지 않을 코드라 이부분 리뷰는 생략하겠습니다.

imnotmoon commented on src/components/commons/Footer/Footer.tsx:
```diff
@@ -0,0 +1,7 @@
+  return <div>footer</div>;
+};
+
+export default Footer;
```
default export 🙅

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import { FlexListItem, FlexListItemProps, FlexBoxCreator } from '../../layout/FlexBox';
+
+interface NavItemData {
```
interface도 쓰고 위에서는 type도 썼네요.
어떤 의도로 구분하신건지 궁금해요.

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -0,0 +1,83 @@
+  url: string;
+}
+
+const NAV_ITEM_DATAS: Array<NavItemData> = [
```
👍

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -0,0 +1,83 @@
+];
+
+const Header = () => {
+  return (
```
shorthand return이 적용되지 않고 있네요. 관련된 prettier 설정을 찾아서 적용해주세요.

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -0,0 +1,83 @@
+  color: inherit;
+  box-shadow: 0 2px 10px rgba(21, 21, 69, 0.2);
+  padding: 0 2rem;
+  top: 0;
```
레이아웃 관련이 아닌, 스타일 관련 코드는 삭제해주세요.
단, Header 같은 컴포넌트는 padding을 root에 넣어도 괜찮아요.

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -0,0 +1,83 @@
+`;
+
+const HeaderNav = styled.nav`
+  display: grid;
+  grid-template-columns: auto;
+  grid-template-rows: 1fr;
```
굳이 grid 필요 없을것 같아요. 다른 방법을 고려해주세요.

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -0,0 +1,83 @@
+  cursor: pointer;
+  text-decoration: none;
+  padding: 0 10px;
+  transition: all 400ms ease;
```
스타일 관련된 코드 삭제해주세요.
이걸 왜 자꾸 하라고 하나면 이후에 디자인이 나왔을때 지우는것조차 비용이 될 수 있어요.
이 코드가 꼭 필요한건지 / 미처 팔로업 하지 못한 디자인 스펙이 있는건지 확인하는 비용이 생겨요.

imnotmoon commented on src/components/commons/Icon/ICON_SET.ts:
```diff
@@ -0,0 +1,25 @@
+import { ViewBoxSize } from '.';
+
+//icon data types
+export type IconName = 'arrowLeft' | 'heart' | 'arrowRight';
```
keyof ICON_SET

imnotmoon commented on src/components/commons/Icon/ICON_SET.ts:
```diff
@@ -0,0 +1,25 @@
+  [key: string]: any;
+}
+
+export const ICON_SET: Record<IconName, IconData> = {
```
이런건 처음 보는데 좋은 처리같네요 👍

imnotmoon commented on src/components/commons/Icon/Icon.tsx:
```diff
@@ -0,0 +1,44 @@
+  color?: string;
+}
+
+const SVGIcon = ({ size, viewBoxSize, iconName, color, ...args }: IconProps) => {
```
```suggestion
const SVGIcon = ({ size, viewBoxSize, iconName, color, ...props }: IconProps) => {
```

imnotmoon commented on src/components/commons/Icon/Icon.tsx:
```diff
@@ -0,0 +1,44 @@
+} from '.';
+import styled from '@emotion/styled';
+
+export interface IconProps {
```
interface 쓰지 맙시다. 구려요.

imnotmoon commented on src/components/commons/Icon/Icon.tsx:
```diff
@@ -0,0 +1,44 @@
+export interface IconProps {
+  iconName: IconName;
+  size?: ViewPortSize;
+  viewBoxSize?: ViewBoxSize;
```
height, width가 다른 viewBox가 필요한 경우엔 어떻게 대응하죠?

imnotmoon commented on src/components/commons/Icon/Icon.tsx:
```diff
@@ -0,0 +1,44 @@
+
+const SVGIcon = ({ size, viewBoxSize, iconName, color, ...args }: IconProps) => {
+  //viewPortSize
+  const {
+    minX,
+    minY,
+    width: viewboxWidth,
+    height: viewboxHeight,
+  } = useMemo(() => setViewBoxSize(viewBoxSize, iconName), [viewBoxSize, iconName]);
+  //viewBoxSize
+  const { width: iconWidth, height: iconHeight } = useMemo(() => setViewPortSize(size), [size]);
+  // path to be drawn
+  const icon_data = useMemo(() => getIconData(iconName), [iconName]);
```
1. 이렇게 고도화된 처리가 필요한 아이콘이 있을지 궁금합니다. 저는 이런걸 경험해보지 못했어요.
2. path를 따로 처리하는것도 과한 처리같아요. [Compound component](https://www.smashingmagazine.com/2021/08/compound-components-react/)를 참고해보세요. SvgIcon 컴포넌트는 간단한 스타일이나 어떤 아이콘을 쓸건지 정도만 처리해주고 다른 책임은 사용처에 위임하는게 좋아요. 함수는 하나의 일만 잘하면 됩니다.
3. 갑자기 스네이크 케이스가 나왔네요. 코드 정리가 덜 된것 같아요.

imnotmoon commented on src/components/commons/Icon/icon.util.ts:
```diff
@@ -0,0 +1,40 @@
+const DEFAULT_ICON_COLOR = 'black';
+
+// IconSize(ViewPort Size)를 결정해주는 함수, 사용자가 단일 number로 사이즈를 입력할 수 있도록 타입체킹
+export const setViewPortSize = (size: ViewPortSize): ViewPortObject => {
```
여기도 너무 과하게 똑똑해지려다보니 필요 없는 코드가 많이 생긴것 같아요.
내일 더 이야기해보고 살릴지 죽일지 결정할게요.

imnotmoon commented on src/components/commons/Icon/icon.util.ts:
```diff
@@ -0,0 +1,40 @@
+  }
+  //ViewBoxString
+  if (typeof IconViewBoxSize === 'string') {
+    let parsed;
+    parsed = IconViewBoxSize.split(' ').map((val) => Number(val));
```
🙅

imnotmoon commented on src/components/layout/FlexBox/flex.util.ts:
```diff
@@ -0,0 +1,117 @@
+const DEFAULT_SUB_AXIS_ITEM_POS_ONE_LINE = 'center';
+const DEFAULT_SUB_AXIS_ITEM_POS = 'stretch';
+
+class FlexBox {
```
여기는 필요한 코드가 맞는것 같아요.
- 디자인이 들어가지 않음
- 개선의 여지는 있지만 사용하기 너무 어렵지도 않음

다만 class 기반 코드라 조금 거부감이 들긴 하네요. 굳이 클래스로 만들어서 싱글톤으로 운용할 필요가 있나? 하는 생각이 들어요.
함수 컴포넌트 + prop으로 쓰면 더 편리할것 같은데 리팩토링 고려해봐주세요.

imnotmoon commented on src/pages/_app.tsx:
```diff
@@ -1,5 +1,27 @@
 
 export default function App({ Component, pageProps }: AppProps) {
-  return <Component {...pageProps} />;
+  const [queryClient] = useState(() => new QueryClient());
```
queryClient는 state로 관리하면 안됩니다. 컴포넌트 바깥에서 선언하고 필요한 경우엔 useQueryClient를 써주세요.

imnotmoon commented on src/pages/_app.tsx:
```diff
@@ -1,5 +1,27 @@
+  const [queryClient] = useState(() => new QueryClient());
+
+  return (
+    <>
+      <QueryClientProvider client={queryClient}>
+        <Hydrate state={pageProps.dehydratedState}>
+          <Header />
+          <Component {...pageProps} />
+          <Footer />
+        </Hydrate>
+        <ReactQueryDevtools />
+      </QueryClientProvider>
+    </>
```
```suggestion
      <QueryClientProvider client={queryClient}>
        <Hydrate state={pageProps.dehydratedState}>
          <Header />
          <Component {...pageProps} />
          <Footer />
        </Hydrate>
        <ReactQueryDevtools />
      </QueryClientProvider>
```

imnotmoon commented on src/pages/board/post/[id]/index.tsx:
```diff
@@ -0,0 +1,20 @@
+
+export default PostPage;
+
+export const getServerSideProps: GetServerSideProps = () => {};
```
```suggestion
```

imnotmoon commented on src/pages/board/post/[id]/index.tsx:
```diff
@@ -0,0 +1,20 @@
+  const id = router.query.id;
+
+  return (
+    <div>
+      <PostList></PostList>
+      <PostListNav></PostListNav>
+    </div>
```
지금 상황에선 딱 이정도 코드가 적당합니다.
👍

imnotmoon commented on src/styles/common.scss:
```diff
@@ -0,0 +1,46 @@
+* {
```
여기도 전부 확인해서 불필요한 코드 제거해주세요.

imnotmoon replied:
sass가 여기서 필요했군요.
근데 이거 하나 쓰려고 의존성 추가할바엔 그냥 pure css로 쓰는건 어떨까 싶기도 하네요.

Rengod95 replied:
저도 동의하는 부분입니다. scss 종속성을 제거하고 순수 css 로 변경하겠습니다!

imnotmoon commented on src/styles/common.scss:
```diff
@@ -0,0 +1,46 @@
+  text-decoration:none;
+  list-style: none;
+  box-sizing: border-box;
+  &::-webkit-scrollbar {
```
브라우저의 기본 동작/디자인은 모든 유저가 '당연히' 예상하는 것들입니다.
꼭 없애야 하는게 아니면 유지해주세요.

Rengod95 replied:
box-sizing 부분을 제외하고 전부 수정했습니다!

imnotmoon commented on src/styles/config.scss:
```diff
@@ -0,0 +1,17 @@
+// 프로젝트 임시 컬러 프로필
```
여기도 확인해서 필요 없어요. 삭제해주세요.

imnotmoon replied:
[23994ef](https://github.com/Rengod95/Dynamic_Roadmap/pull/3/commits/23994ef5e5ddbd9d47d8cf3450021f353f2e681e)

Rengod95 committed:
Update src/components/commons/Button/Button.tsx

Co-authored-by: Sanghyeok Moon <53335940+imnotmoon@users.noreply.github.com>

Rengod95 committed:
feat: PR 코드리뷰 반영

- 컴포넌트 별 파일 구조 분리
 - 예로 Button.tsx 에 대해
 - Button.css.ts
 - Button.util.ts 혹은 Button.hooks.ts
 - index.ts

 - PostList.tsx 로직 분리
  - 데이터 패칭 로직을 훅으로 분리했습니다.

- 필요없는 scss 파일 삭제
 - global css를 위한 common.css 를 제외하고 삭제했습니다.
 - 종속성으로 인한 merge conflict가 발생하지 않도록 일단 sass 종속은 지우지 않았습니다.

 - Styled Component Naming Convention 은 다음 PR로 반영하겠습니다.

Rengod95 committed:
Merge branch 'feature/SetDefaultUI' of https://github.com/Rengod95/Dynamic_Roadmap into feature/SetDefaultUI

imnotmoon reviewed (approved):
수정사항이 너무 많아서 다시 보기 어렵습니다.
일단은 머지해주세요.

Rengod95 committed:
Merge branch 'main' of https://github.com/Rengod95/Dynamic_Roadmap into feature/SetDefaultUI
